### PR TITLE
Add backfill ORM models

### PR DESCRIPTION
**Pull request by dstandish**
This is one of the first steps to implement AIP-78.

The structure of the models may change over time (1) as new data awareness capabilities or (2) as dictated by implementation in the scheduler.

### CODE REVIEW

File: airflow/migrations/versions/0032_3_0_0_add_tables_for_backfill.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+revision = "522625f6d606"
+down_revision = "1cdc775ca98f"
+branch_labels = None
+depends_on = None
+airflow_version = "3.0.0"
+
+
+def upgrade():
+    """Apply Add tables for backfill."""
+    # ### commands auto generated by Alembic - please adjust! ###
```
Comment (ephraimbuddy) [inline airflow/migrations/versions/0032_3_0_0_add_tables_for_backfill.py]:
```suggestion
```
We can remove this and some other ones like it

File: airflow/migrations/versions/0032_3_0_0_add_tables_for_backfill.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        sa.Column("created_at", airflow.utils.sqlalchemy.UtcDateTime(timezone=True), nullable=False),
+        sa.Column("completed_at", airflow.utils.sqlalchemy.UtcDateTime(timezone=True), nullable=True),
+        sa.Column("updated_at", airflow.utils.sqlalchemy.UtcDateTime(timezone=True), nullable=False),
+        sa.PrimaryKeyConstraint("id", name=op.f("backfill_pkey")),
+    )
+    op.create_table(
+        "backfill_dag_run",
+        sa.Column("id", sa.Integer(), autoincrement=True, nullable=False),
+        sa.Column("backfill_id", sa.Integer(), nullable=False),
+        sa.Column("dag_run_id", sa.Integer(), nullable=True),
```
Comment (ephraimbuddy) [inline airflow/migrations/versions/0032_3_0_0_add_tables_for_backfill.py:64]: Does this mean we can have a backfill without a dagrun or should we make it nullable=False?

File: airflow/models/__init__.py [hunk 3]
```diff
@@ -23,6 +23,8 @@
 __all__ = [
     "DAG",
     "ID_LEN",
+    "Backfill",
+    "BackfillDagRun",
```
Comment (ephraimbuddy) [inline airflow/models/__init__.py]: There's an instruction not to add new models to this at line 22 above..

File: airflow/migrations/versions/0032_3_0_0_add_tables_for_backfill.py (hunk 2 above)
Comment (dstandish) [reply airflow/migrations/versions/0032_3_0_0_add_tables_for_backfill.py:64]: the intention is to be able to record what happens when a particular backfill request has a range that includes a dag run that already exists.  in that case we could create this mapping record but indicate that no run was created and why.  that's the thought anyway.

File: airflow/models/__init__.py (hunk 3 above)
Comment (dstandish) [reply airflow/models/__init__.py]: does "this" mean "anywhere in this module" or just in the "__all__"?

Comment (dstandish) [reply airflow/models/__init__.py]: removed

File: airflow/migrations/versions/0032_3_0_0_add_tables_for_backfill.py (hunk 1 above)
Comment (dstandish) [reply airflow/migrations/versions/0032_3_0_0_add_tables_for_backfill.py]: done

File: airflow/models/__init__.py [hunk 4]
```diff
@@ -86,6 +87,8 @@ def __getattr__(name):
 __lazy_imports = {
     "DAG": "airflow.models.dag",
     "ID_LEN": "airflow.models.base",
+    "Backfill": "airflow.models.backfill",
+    "BackfillDagRun": "airflow.models.backfill",
```
Comment (jedcunningham) [inline airflow/models/__init__.py]: My 2c, let's not even add these here.

File: airflow/models/backfill.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+from airflow.utils.sqlalchemy import UtcDateTime
+
+
+class Backfill(Base):
+    """Model representing a backfill job."""
+
+    __tablename__ = "backfill"
+
+    id = Column(Integer, primary_key=True, autoincrement=True)
+    dag_id = Column(String(ID_LEN, **COLLATION_ARGS))
```
Comment (jedcunningham) [inline airflow/models/backfill.py]:
```suggestion
    dag_id = Column(String(ID_LEN, **COLLATION_ARGS))
```

Shouldn't we use [StringID](https://github.com/apache/airflow/blob/b599d81121f3b9103658fba9c33d6f3bebe33cf6/airflow/models/base.py#L84) here instead?

File: airflow/models/__init__.py (hunk 4 above)
Comment (dstandish) [reply airflow/models/__init__.py]: IIRC i needed to add them _somewhere_ in the module for it to work.
i'll take a look.

Comment (jedcunningham) [reply airflow/models/__init__.py]: I think it's needed in `import_all_models`, not here :)

Comment (dstandish) [reply airflow/models/__init__.py]: ok done thanks

File: airflow/models/backfill.py (hunk 5 above)
Comment (dstandish) [reply airflow/models/backfill.py]: sure, done

**Review by jedcunningham — APPROVED** [review]

**Review by vincbeck — APPROVED** [review]

**Comment by dstandish** [comment]
closes https://github.com/apache/airflow/issues/42331

**Pull request merged**